Support Enum with self defined members

### Current problem

False positive E1101 are reported for self defined member in Enums. 

Example:
- [Enum `Layout` w/ self defined members `col` & `row`](https://github.com/OldIMP/volksschule/blob/ba0dc09ad4590439d5ae8c3c7b7639b45869db70/r_quiz.py#L139)
- [Here](https://github.com/OldIMP/volksschule/blob/ba0dc09ad4590439d5ae8c3c7b7639b45869db70/r_quiz.py#L174) will then be reported falsely as E1101:
  > r_quiz.py:174:12: E1101: Instance of 'SIMPLE' has no '`row`' member (no-member)
  > r_quiz.py:174:12: E1101: Instance of 'SCHRIFTLICH' has no '`row`' member (no-member)
  > r_quiz.py:174:25: E1101: Instance of 'SIMPLE' has no '`col`' member (no-member)
  > r_quiz.py:174:25: E1101: Instance of 'SCHRIFTLICH' has no '`col`' member (no-member)

#### pylint --version
pylint 2.11.1
astroid 2.8.2
Python 3.9.7 | packaged by conda-forge | (default, Sep 29 2021, 19:15:42) [MSC v.1916 64 bit (AMD64)]

### Desired solution

According to [Python doc](https://docs.python.org/3/library/enum.html#planet), self defined members in Enums are allowed. Therefore no issues should be reported for the example above. 

### Additional context

_No response_

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -439,6 +439,7 @@ contributors:
 - Jacques Kvam <jwkvam@gmail.com>
 - Jace Browning <jacebrowning@gmail.com>: updated default report format with clickable paths
 - JT Olds <jtolds@xnet5.com>
+- Huw Jones <huw@huwcbjones.co.uk> (huwcbjones)
 - Hayden Richards <62866982+SupImDos@users.noreply.github.com> (SupImDos)
   * Fixed "no-self-use" for async methods
   * Fixed "docparams" extension for async functions and methods

--- a/ChangeLog
+++ b/ChangeLog
@@ -249,6 +249,10 @@ Release date: TBA
 
   Closes #5219
 
+* Fixed false positive ``no-member`` for Enums with self-defined members.
+
+  Closes #5138
+
 * Fix false negative for ``no-member`` when attempting to assign an instance
   attribute to itself without any prior assignment.
 

--- a/doc/whatsnew/2.14.rst
+++ b/doc/whatsnew/2.14.rst
@@ -260,6 +260,10 @@ Other Changes
 
   Closes #5219
 
+* Fixed false positive ``no-member`` for Enums with self-defined members.
+
+  Closes #5138
+
 * Fix false negative for ``no-member`` when attempting to assign an instance
   attribute to itself without any prior assignment.
 

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -526,22 +526,22 @@ def _emit_no_member(
 
 
 def _get_all_attribute_assignments(
-    node: astroid.FunctionDef, name: str | None = None
+    node: nodes.FunctionDef, name: str | None = None
 ) -> set[str]:
     attributes: set[str] = set()
-    for child in node.nodes_of_class((astroid.Assign, astroid.AnnAssign)):
+    for child in node.nodes_of_class((nodes.Assign, nodes.AnnAssign)):
         targets = []
-        if isinstance(child, astroid.Assign):
+        if isinstance(child, nodes.Assign):
             targets = child.targets
-        elif isinstance(child, astroid.AnnAssign):
+        elif isinstance(child, nodes.AnnAssign):
             targets = [child.target]
         for assign_target in targets:
-            if isinstance(assign_target, astroid.Tuple):
+            if isinstance(assign_target, nodes.Tuple):
                 targets.extend(assign_target.elts)
                 continue
             if (
-                isinstance(assign_target, astroid.AssignAttr)
-                and isinstance(assign_target.expr, astroid.Name)
+                isinstance(assign_target, nodes.AssignAttr)
+                and isinstance(assign_target.expr, nodes.Name)
                 and (name is None or assign_target.expr.name == name)
             ):
                 attributes.add(assign_target.attrname)
@@ -556,16 +556,13 @@ def _enum_has_attribute(
             (b.parent for b in owner.bases if isinstance(b.parent, nodes.ClassDef)),
             None,
         )
-        if enum_def is None:
-            return False
+        assert enum_def is not None, "enum_def unexpectedly None"
     else:
         enum_def = owner
 
     # Find __new__ and __init__
     dunder_new = next((m for m in enum_def.methods() if m.name == "__new__"), None)
     dunder_init = next((m for m in enum_def.methods() if m.name == "__init__"), None)
-    if dunder_new is None and dunder_init is None:
-        return False
 
     enum_attributes: set[str] = set()
 
@@ -573,11 +570,7 @@ def _enum_has_attribute(
     if dunder_new:
         # Get the object returned in __new__
         returned_obj_name = next(
-            (
-                c.value
-                for c in dunder_new.get_children()
-                if isinstance(c, astroid.Return)
-            ),
+            (c.value for c in dunder_new.get_children() if isinstance(c, nodes.Return)),
             None,
         )
         if returned_obj_name is not None:

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -525,6 +525,21 @@ def _emit_no_member(
     return True
 
 
+def _get_all_attribute_assignments(
+    node: astroid.FunctionDef, name: str | None = None
+) -> set[str]:
+    attributes = set()
+    for child in node.nodes_of_class(astroid.Assign):
+        for assign_target in child.targets:
+            if not isinstance(assign_target, astroid.AssignAttr):
+                continue
+            if not isinstance(assign_target.expr, astroid.Name):
+                continue
+            if name is None or assign_target.expr.name == name:
+                attributes.add(assign_target.attrname)
+    return attributes
+
+
 def _enum_has_attribute(
     owner: Union[astroid.Instance, nodes.ClassDef], node: nodes.Attribute
 ) -> bool:
@@ -533,31 +548,38 @@ def _enum_has_attribute(
     while not isinstance(enum_def, astroid.ClassDef):
         enum_def = enum_def.parent
 
-    # Find __new__
+    # Find __new__ and __init__
     dunder_new = next((m for m in enum_def.methods() if m.name == "__new__"), None)
-    if dunder_new is None:
-        return False
-
-    # Get the returned object
-    returned_obj_name = next(
-        (c.value for c in dunder_new.get_children() if isinstance(c, astroid.Return)),
-        None,
-    )
-    if returned_obj_name is None:
+    dunder_init = next((m for m in enum_def.methods() if m.name == "__init__"), None)
+    if dunder_new is None and dunder_init is None:
         return False
 
     enum_attributes = set()
-    # Iterate over children and find Enum attr assignments
-    for child in dunder_new.get_children():
-        if not isinstance(child, astroid.Assign):
-            continue
-        for assign_target in child.targets:
-            if not isinstance(assign_target, astroid.AssignAttr):
-                continue
-            if not isinstance(assign_target.expr, astroid.Name):
-                continue
-            if assign_target.expr.name == returned_obj_name.name:
-                enum_attributes.add(assign_target.attrname)
+
+    # Find attributes defined in __new__
+    if dunder_new:
+        # Get the object returned in __new__
+        returned_obj_name = next(
+            (
+                c.value
+                for c in dunder_new.get_children()
+                if isinstance(c, astroid.Return)
+            ),
+            None,
+        )
+        if returned_obj_name is not None:
+            # Find all attribute assignments to the returned object
+            enum_attributes |= _get_all_attribute_assignments(
+                dunder_new, returned_obj_name.name
+            )
+
+    # Find attributes defined in __init__
+    if dunder_init and dunder_init.body and dunder_init.args:
+        # Grab the name referring to `self` from the function def
+        enum_attributes |= _get_all_attribute_assignments(
+            dunder_init, dunder_init.args.arguments[0].name
+        )
+
     return node.attrname in enum_attributes
 
 

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -451,6 +451,8 @@ def _emit_no_member(
                 return False
             if metaclass:
                 # Renamed in Python 3.10 to `EnumType`
+                if _enum_has_attribute(owner, node):
+                    return False
                 return metaclass.qname() in {"enum.EnumMeta", "enum.EnumType"}
             return False
         if not has_known_bases(owner):
@@ -521,6 +523,40 @@ def _emit_no_member(
         node_origin, parent = parent, parent.parent
 
     return True
+
+
+def _enum_has_attribute(owner, node) -> bool:
+    enum_def = owner
+    # Traverse the AST to find the Enum ClassDef
+    while not isinstance(enum_def, astroid.ClassDef):
+        enum_def = enum_def.parent
+
+    # Find __new__
+    dunder_new = next((m for m in enum_def.methods() if m.name == "__new__"), None)
+    if dunder_new is None:
+        return False
+
+    # Get the returned object
+    returned_obj_name = next(
+        (c.value for c in dunder_new.get_children() if isinstance(c, astroid.Return)),
+        None,
+    )
+    if returned_obj_name is None:
+        return False
+
+    enum_attributes = set()
+    # Iterate over children and find Enum attr assignments
+    for child in dunder_new.get_children():
+        if not isinstance(child, astroid.Assign):
+            continue
+        for assign_target in child.targets:
+            if not isinstance(assign_target, astroid.AssignAttr):
+                continue
+            if not isinstance(assign_target.expr, astroid.Name):
+                continue
+            if assign_target.expr.name == returned_obj_name.name:
+                enum_attributes.add(assign_target.attrname)
+    return node.attrname in enum_attributes
 
 
 def _determine_callable(

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -528,7 +528,7 @@ def _emit_no_member(
 def _get_all_attribute_assignments(
     node: astroid.FunctionDef, name: str | None = None
 ) -> set[str]:
-    attributes = set()
+    attributes: set[str] = set()
     for child in node.nodes_of_class((astroid.Assign, astroid.AnnAssign)):
         targets = []
         if isinstance(child, astroid.Assign):
@@ -536,11 +536,14 @@ def _get_all_attribute_assignments(
         elif isinstance(child, astroid.AnnAssign):
             targets = [child.target]
         for assign_target in targets:
-            if not isinstance(assign_target, astroid.AssignAttr):
+            if isinstance(assign_target, astroid.Tuple):
+                targets.extend(assign_target.elts)
                 continue
-            if not isinstance(assign_target.expr, astroid.Name):
-                continue
-            if name is None or assign_target.expr.name == name:
+            if (
+                isinstance(assign_target, astroid.AssignAttr)
+                and isinstance(assign_target.expr, astroid.Name)
+                and (name is None or assign_target.expr.name == name)
+            ):
                 attributes.add(assign_target.attrname)
     return attributes
 
@@ -557,9 +560,6 @@ def _enum_has_attribute(
             return False
     else:
         enum_def = owner
-    # Traverse the AST to find the Enum ClassDef
-    while not isinstance(enum_def, astroid.ClassDef):
-        enum_def = enum_def.parent
 
     # Find __new__ and __init__
     dunder_new = next((m for m in enum_def.methods() if m.name == "__new__"), None)
@@ -567,7 +567,7 @@ def _enum_has_attribute(
     if dunder_new is None and dunder_init is None:
         return False
 
-    enum_attributes = set()
+    enum_attributes: set[str] = set()
 
     # Find attributes defined in __new__
     if dunder_new:

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -543,7 +543,15 @@ def _get_all_attribute_assignments(
 def _enum_has_attribute(
     owner: Union[astroid.Instance, nodes.ClassDef], node: nodes.Attribute
 ) -> bool:
-    enum_def = owner
+    if isinstance(owner, astroid.Instance):
+        enum_def = next(
+            (b.parent for b in owner.bases if isinstance(b.parent, nodes.ClassDef)),
+            None,
+        )
+        if enum_def is None:
+            return False
+    else:
+        enum_def = owner
     # Traverse the AST to find the Enum ClassDef
     while not isinstance(enum_def, astroid.ClassDef):
         enum_def = enum_def.parent

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -556,6 +556,15 @@ def _enum_has_attribute(
             (b.parent for b in owner.bases if isinstance(b.parent, nodes.ClassDef)),
             None,
         )
+
+        if enum_def is None:
+            # We don't inherit from anything, so try to find the parent
+            # class definition and roll with that
+            enum_def = node
+            while enum_def is not None and not isinstance(enum_def, nodes.ClassDef):
+                enum_def = enum_def.parent
+
+        # If this blows, something is clearly wrong
         assert enum_def is not None, "enum_def unexpectedly None"
     else:
         enum_def = owner

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -529,8 +529,13 @@ def _get_all_attribute_assignments(
     node: astroid.FunctionDef, name: str | None = None
 ) -> set[str]:
     attributes = set()
-    for child in node.nodes_of_class(astroid.Assign):
-        for assign_target in child.targets:
+    for child in node.nodes_of_class((astroid.Assign, astroid.AnnAssign)):
+        targets = []
+        if isinstance(child, astroid.Assign):
+            targets = child.targets
+        elif isinstance(child, astroid.AnnAssign):
+            targets = [child.target]
+        for assign_target in targets:
             if not isinstance(assign_target, astroid.AssignAttr):
                 continue
             if not isinstance(assign_target.expr, astroid.Name):
@@ -541,7 +546,7 @@ def _get_all_attribute_assignments(
 
 
 def _enum_has_attribute(
-    owner: Union[astroid.Instance, nodes.ClassDef], node: nodes.Attribute
+    owner: astroid.Instance | nodes.ClassDef, node: nodes.Attribute
 ) -> bool:
     if isinstance(owner, astroid.Instance):
         enum_def = next(

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -450,9 +450,9 @@ def _emit_no_member(
             except astroid.MroError:
                 return False
             if metaclass:
-                # Renamed in Python 3.10 to `EnumType`
                 if _enum_has_attribute(owner, node):
                     return False
+                # Renamed in Python 3.10 to `EnumType`
                 return metaclass.qname() in {"enum.EnumMeta", "enum.EnumType"}
             return False
         if not has_known_bases(owner):
@@ -525,7 +525,9 @@ def _emit_no_member(
     return True
 
 
-def _enum_has_attribute(owner, node) -> bool:
+def _enum_has_attribute(
+    owner: Union[astroid.Instance, nodes.ClassDef], node: nodes.Attribute
+) -> bool:
     enum_def = owner
     # Traverse the AST to find the Enum ClassDef
     while not isinstance(enum_def, astroid.ClassDef):

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -1,0 +1,22 @@
+# [missing-module-docstring]
+from enum import IntEnum
+
+
+class Day(IntEnum):  # [missing-class-docstring]
+    MONDAY = (1, "Mon")
+    TUESDAY = (2, "Tue")
+    WEDNESDAY = (3, "Wed")
+    THURSDAY = (4, "Thu")
+    FRIDAY = (5, "Fri")
+    SATURDAY = (6, "Sat")
+    SUNDAY = (7, "Sun")
+
+    def __new__(cls, value, abbr):
+        obj = int.__new__(cls, value)
+        obj._value_ = value
+        obj.abbr = abbr
+        return obj
+
+
+print(Day.FRIDAY.abbr)
+print(Day.FRIDAY.foo)  # [no-member]

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -20,6 +20,9 @@ class Day(IntEnum):
             obj.abbr = ""
         return obj
 
+    def __repr__(self):
+        return f"{self._value_}: {self.foo}"  # [no-member]
+
 
 print(Day.FRIDAY.abbr)
 print(Day.FRIDAY.foo)  # [no-member]

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -1,5 +1,5 @@
 # pylint: disable=missing-docstring
-from enum import IntEnum
+from enum import IntEnum, Enum
 
 
 class Day(IntEnum):
@@ -11,12 +11,32 @@ class Day(IntEnum):
     SATURDAY = (6, "Sat")
     SUNDAY = (7, "Sun")
 
-    def __new__(cls, value, abbr):
+    def __new__(cls, value, abbr=None):
         obj = int.__new__(cls, value)
         obj._value_ = value
-        obj.abbr = abbr
+        if abbr:
+            obj.abbr = abbr
+        else:
+            obj.abbr = ""
         return obj
 
 
 print(Day.FRIDAY.abbr)
 print(Day.FRIDAY.foo)  # [no-member]
+
+
+class Length(Enum):
+    METRE = "metre", "m"
+    MILE = "mile", "m", True
+
+    def __init__(self, text: str, unit: str,  is_imperial: bool = False):
+        self.text = text
+        self.unit = unit
+        if is_imperial:
+            self.suffix = " (imp)"
+        else:
+            self.suffix = ""
+
+
+print(f"100 {Length.METRE.unit}{Length.METRE.suffix}")
+print(Length.MILE.foo)  # [no-member]

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -1,3 +1,4 @@
+"""Tests for self-defined Enum members (https://github.com/PyCQA/pylint/issues/5138)"""
 # pylint: disable=missing-docstring
 from enum import IntEnum, Enum
 
@@ -43,3 +44,20 @@ class Length(Enum):
 
 print(f"100 {Length.METRE.unit}{Length.METRE.suffix}")
 print(Length.MILE.foo)  # [no-member]
+
+
+class Binary(int, Enum):
+    ZERO = 0
+    ONE = 1
+
+    def __init__(self, value: int) -> None:
+        super().__init__()
+        self.str, self.inverse = str(value), abs(value - 1)
+
+        def no_op(_value):
+            pass
+        value_squared = value ** 2
+        no_op(value_squared)
+
+
+print(f"1={Binary.ONE.value} (Inverted: {Binary.ONE.inverse}")

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -1,8 +1,8 @@
-# [missing-module-docstring]
+# pylint: disable=missing-docstring
 from enum import IntEnum
 
 
-class Day(IntEnum):  # [missing-class-docstring]
+class Day(IntEnum):
     MONDAY = (1, "Mon")
     TUESDAY = (2, "Tue")
     WEDNESDAY = (3, "Wed")

--- a/tests/functional/e/enum_self_defined_member_5138.py
+++ b/tests/functional/e/enum_self_defined_member_5138.py
@@ -33,8 +33,8 @@ class Length(Enum):
     MILE = "mile", "m", True
 
     def __init__(self, text: str, unit: str,  is_imperial: bool = False):
-        self.text = text
-        self.unit = unit
+        self.text: str = text
+        self.unit: str = unit
         if is_imperial:
             self.suffix = " (imp)"
         else:

--- a/tests/functional/e/enum_self_defined_member_5138.txt
+++ b/tests/functional/e/enum_self_defined_member_5138.txt
@@ -1,2 +1,3 @@
-no-member:25:6:25:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
-no-member:42:6:42:21::Instance of 'MILE' has no 'foo' member:INFERENCE
+no-member:24:34:24:42:Day.__repr__:Instance of 'Day' has no 'foo' member:INFERENCE
+no-member:28:6:28:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
+no-member:45:6:45:21::Instance of 'MILE' has no 'foo' member:INFERENCE

--- a/tests/functional/e/enum_self_defined_member_5138.txt
+++ b/tests/functional/e/enum_self_defined_member_5138.txt
@@ -1,1 +1,2 @@
-no-member:22:6:22:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
+no-member:25:6:25:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
+no-member:42:6:42:21::Instance of 'MILE' has no 'foo' member:INFERENCE

--- a/tests/functional/e/enum_self_defined_member_5138.txt
+++ b/tests/functional/e/enum_self_defined_member_5138.txt
@@ -1,3 +1,1 @@
-missing-module-docstring:1:0:None:None::Missing module docstring:HIGH
-missing-class-docstring:5:0:18:18:Day:Missing class docstring:HIGH
 no-member:22:6:22:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE

--- a/tests/functional/e/enum_self_defined_member_5138.txt
+++ b/tests/functional/e/enum_self_defined_member_5138.txt
@@ -1,3 +1,3 @@
-no-member:24:34:24:42:Day.__repr__:Instance of 'Day' has no 'foo' member:INFERENCE
-no-member:28:6:28:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
-no-member:45:6:45:21::Instance of 'MILE' has no 'foo' member:INFERENCE
+no-member:25:34:25:42:Day.__repr__:Instance of 'Day' has no 'foo' member:INFERENCE
+no-member:29:6:29:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE
+no-member:46:6:46:21::Instance of 'MILE' has no 'foo' member:INFERENCE

--- a/tests/functional/e/enum_self_defined_member_5138.txt
+++ b/tests/functional/e/enum_self_defined_member_5138.txt
@@ -1,0 +1,3 @@
+missing-module-docstring:1:0:None:None::Missing module docstring:HIGH
+missing-class-docstring:5:0:18:18:Day:Missing class docstring:HIGH
+no-member:22:6:22:20::Instance of 'FRIDAY' has no 'foo' member:INFERENCE

